### DISCUSSION TITLE
How to avoid runtime conflict?

### DISCUSSION THREAD

**Discussion by kwonoj** [wasmerio/wasmer#3966] [Q&A]
Stack:

> Panic: PanicInfo { payload: Any { .. }, message: Some(Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.), location: Location { file: "/Users/ojkwon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmer-wasix-0.4.0/src/runtime/task_manager/mod.rs", line: 104, col: 24 }, can_unwind: true }

I'm trying to integrate existing runner logics into different implementation. The existing runtime seems having an initialized tokio runtime, which looks like causing a conflict when wasix trying to do the same thing.

Am I understanding the issue correctly? and is there way to avoid / workaround this?

**Comment by kwonoj** [comment]
Reproducible PR: https://github.com/vercel/next.js/pull/50401

**Comment by Michael-F-Bryan** [comment]
You can make sure WASIX and your calling code share the same Tokio runtime by passing `tokio::runtime::Handle::current()` to `TokioTaskManager::new()` when setting up your WASI runtime.

```rust
let tasks = TokioTaskManager::new(Handle::current());
let rt = PluggableRuntime::new(Arc::new(tasks));
```

We've got a bunch of integration tests that run with `#[tokio::test]` and this is how we set the WASI runtime up to share the Tokio runtime.

https://github.com/wasmerio/wasmer/blob/c2867bb0167efa05f4fa6dff91a1b9e2a750ae14/lib/wasi/tests/runners.rs#L218-L229

That avoids most of the *"Cannot start a runtime from within a runtime"* panics.

**Reply by kwonoj** [reply]
I tried suggestion, however still gettting the same unfortunately.

Currently runtime's constructed as https://github.com/swc-project/swc/blob/main/crates/swc_plugin_runner/src/wasix_runtime.rs#L23-L38, and getting same panics.

**Reply by kwonoj** [reply]
Update: for me, calling `WasiFunctionEnv::cleanup` causes this regardless of setting shared handle or not.
